zaza/utilities/generic.py

```diff
@@ -165,3 +166,251 @@ def get_yaml_config(config_file):
+    # before upgrade. Finally, for some applications this is aribtrary but
+    # generalized.
+    leader = None
+    non_leaders = []
+    for unit in status["units"]:
+        if status["units"][unit].get("leader"):
+            leader = unit
+        else:
+            non_leaders.append(unit)
+
+    # Pause the non-leaders
+    for unit in non_leaders:
+        if pause_non_leader_subordinate:
+            if status["units"][unit].get("subordinates"):
+                for subordinate in status["units"][unit]["subordinates"]:
+                    logging.info("Pausing {}".format(subordinate))
+                    model.run_action(subordinate, "pause", action_params={})
+        if pause_non_leader_primary:
+            logging.info("Pausing {}".format(unit))
+            model.run_action(unit, "pause", action_params={})
+
+    # Series upgrade the leader
+    logging.info("Series upgrade leader: {}".format(leader))
+    series_upgrade(leader, status["units"][leader]["machine"],
+                   from_series=from_series, to_series=to_series,
+                   origin=origin, workaround_script=workaround_script,
+                   files=files)
+
+    # Series upgrade the non-leaders
+    for unit in non_leaders:
+        logging.info("Series upgrade non-leader unit: {}"
+                     .format(unit))
+        series_upgrade(unit, status["units"][unit]["machine"],
+                       from_series=from_series, to_series=to_series,
+                       origin=origin, workaround_script=workaround_script,
+                       files=files)
+
+
+def series_upgrade(unit_name, machine_num,
+                   from_series="trusty", to_series="xenial",
+                   origin='openstack-origin',
+                   files=None, workaround_script=None):
+    """Perform series upgrade on a unit.
+
+    :param unit_name: Unit Name
+    :type unit_name: str
+    :param machine_num: Machine number
+    :type machine_num: str
+    :param from_series: The series from which to upgrade
+    :type from_series: str
+    :param to_series: The series to which to upgrade
+    :type to_series: str
+    :param origin: The configuration setting variable name for changing origin
+                   source. (openstack-origin or source)
+    :type origin: str
+    :param files: Workaround files to scp to unit under upgrade
+    :type files: list
+    :param workaround_script: Workaround script to run during series upgrade
+    :type workaround_script: str
+    :returns: None
+    :rtype: None
+    """
+    logging.info("Series upgrade {}".format(unit_name))
+    application = unit_name.split('/')[0]
+    logging.info("Prepare series upgrade on {}".format(machine_num))
+    juju_utils.prepare_series_upgrade(machine_num, to_series=to_series)
+    logging.info("Watiing for workload status 'unknown' on {}"
+                 .format(unit_name))
+    model.block_until_unit_wl_status(unit_name, "unknown")
```

Comment:
Do we still have charms that have not implemented wl status so would have an 'unknown' status before the series upgrade began? Mongo etc

Reply:
This is a function of series upgrade. After juju upgrade-series prepare $M $S the juju agents stop and report workload status unknonw and agent lost. Here is an example:

Model              Controller  Cloud/Region             Version    SLA          Timestamp
zaza-d4a4dcf30090  upgrades    serverstack/serverstack  2.5-beta1  unsupported  07:42:13-07:00

App              Version          Status   Scale  Charm            Store  Rev  OS      Charm version  Notes
pc-hacluster                      blocked    2/3  hacluster        local    0  ubuntu  f6e9ee9        
percona-cluster  5.5.37-25.10...  blocked    2/3  percona-cluster  local   45  ubuntu  f6e9ee9        

Unit                Workload     Agent  Machine  Public address  Ports     Message
percona-cluster/0   unknown      lost   0        10.5.0.22       3306/tcp  agent lost, see 'juju show-status-log percona-cluster/0'
  pc-hacluster/2    unknown      lost            10.5.0.22                 agent lost, see 'juju show-status-log pc-hacluster/2'
percona-cluster/1*  maintenance  idle   1        10.5.0.12       3306/tcp  Paused. Use 'resume' action to resume normal service.
  pc-hacluster/0*   maintenance  idle            10.5.0.12                 Paused. Use 'resume' action to resume normal service.
percona-cluster/2   maintenance  idle   2        10.5.0.17       3306/tcp  Paused. Use 'resume' action to resume normal service.
  pc-hacluster/1    maintenance  idle            10.5.0.17                 Paused. Use 'resume' action to resume normal service.